cmd/umoci.go

```diff
@@ -188,6 +189,61 @@ func tarUnpack(oci casext.Engine, tag string, bundlePath string, callback layer.
 	return umoci.Unpack(oci, tag, bundlePath, opts)
 }
 
+func which(name string) string {
+	return whichSearch(name, strings.Split(os.Getenv("PATH"), ":"))
+}
+
+func whichSearch(name string, paths []string) string {
+	var search []string
+
+	if strings.ContainsRune(name, os.PathSeparator) {
+		if path.IsAbs(name) {
+			search = []string{name}
+		} else {
+			search = []string{"./" + name}
+		}
+	} else {
+		search = []string{}
+		for _, p := range paths {
+			search = append(search, path.Join(p, name))
+		}
+	}
+
+	for _, fPath := range search {
+		if err := unix.Access(fPath, unix.X_OK); err == nil {
+			return fPath
+		}
+	}
+
+	return ""
+}
+
+func extractSingleSquash(squashFile string, extractDir string) error {
+	err := os.MkdirAll(extractDir, 0755)
+	if err != nil {
+		return err
+	}
+
+	var uCmd []string
+	if config.StorageType == "btrfs" {
```

Comment:
the decision on which tool to use isn't really based on the storageType, but rather the intent of the extract.

Either:
  a.) you are extracting as an archive (you want the .whiteout files present) so that you can overlay mount or maybe modify and re-pack
  b.) you are extracting to use the directory as if it is a layer.

Since we are assuming priviledged, the only difference is '--whiteouts'.

And just to be clear, `unsquashfs` can only be used for 'a' and requires that:
 1. you are privileged and can create device nodes (or squashFile has no device nodes)
 2. `extractDir` is empty or `squashFile` has no whiteout files.

Reply:
But the "intent of the extract" here is exactly storage type: in btrfs we want it to unlink(), and in overlay we want it to mknod.

Reply:
Is that true always?  i was thinking there was different intent between stacker using a layer (and putting it in ~/.stacker or wherever) and a stacker file doing the following, which was what caused me to open #103.

    from:
      type: oci
      url: oci:boot

Reply:
Yes, it is always true. When you see from: type: oci, stacker still extracts the layer using the storage backend so you can run code on it.

Reply:
So there was some code a while ago that probably handled this extraction from tar files incorrectly, https://github.com/opencontainers/umoci/commit/c0f84452c755c30beb2d199121fd096e95656299 and https://github.com/opencontainers/umoci/commit/a425f92e910345405e8c8d8bb67e7da510ac1b65 would have fixed some of that, though. Maybe that's what you were seeing?

Reply:
shrug.  I dont' know.  I thought it was using unsquashfs to extract the oci:boot layer (including the underlying layers) to do the import.  That failed because unsquashfs can't handle .whiteout files.

if its fixed, then good.  If its not, then i'll bother you again sometime.
I suspect my  usecase will be fixed by  use of squashtool here.